src/set_relation/set_relation.cc

```diff
@@ -1642,6 +1642,225 @@ Set* Conjunction::normalize() const {
+    // Replace all uninterpreted function calls with variables
+    // and create a new affine conjunction that is a superset of the
+    // current conjunction.  UFCallMapAndBounds object will
+    // maintain the mapping of temporary variables to UF calls.
+    UFCallMapAndBounds ufcallmap(getTupleDecl());
+    selfcopy->ufCallsToTempVars( ufcallmap );
+
+    Set* constraints = ufcallmap.cloneConstraints();
+
+//std::cout << "constraints = " << constraints->toString() << std::endl;    
+//std::cout << "ufcallmap = " << ufcallmap.toString() << std::endl;
+
+    // update the tuple declaration in selfcopy to match what is
+    // in the constraints
+    selfcopy->setTupleDecl( constraints->getTupleDecl() );
+    
+    // Also need to include all the constraints in selfcopy were UF calls
+    // were replaced by temporaries.
+    Set* selfcopyset = new Set(selfcopy->getTupleDecl());
+    selfcopyset->addConjunction(selfcopy);
+//std::cout << "selfcopyset = " << selfcopyset->toString() << std::endl;
+
+    Set* retval1 = constraints->Intersect(selfcopyset);
+//std::cout << "after intersect = , retval = " << retval1->toString() << std::endl;
+    
+    //////////////////////
+    // (step 2) 
+    // Send the result of step 1 to ISL and back.
+    
+    // (a) get string from result of step 1
+    std::string fromStep1 = retval1->toISLString();
+//std::cout << "fromStep1 = " << fromStep1 << std::endl;
+ 
+    // (b) send through ISL to project out desired tuple variable
+    string fromISL = ISL_project_out(fromStep1, tvar);
+//std::cout<<std::endl<<std::endl<<"from isl:  "<<fromISL<<std::endl<<std::endl;
+
+    // (c) convert back to a Set
+    Set* retval2 = new Set(fromISL);
+//std::cout << "retval2 = " << retval2->toString() << std::endl;
+
+    delete retval1;
+    
+    // Then in (step 3) we will need to use substitute to replace extra
+    // tuple vars with UF calls by making queries to ufcallmap.  Will need
+    // to call setTupleDecl on retval to remove those extra tuple vars that
+    // acted as temporaries to replace uf calls.  Then return retval Set.
+    /////////////////////
+    // (step 3)
+    
+    // FIXME: just assuming one conjunction right now.
+    if (retval2->mConjunctions.size()!=1) {
+        throw assert_exception("Conjunction::normalize: "
+            "currently only handle one Conjunction fromISL");
+    }
+    Conjunction* conj = retval2->mConjunctions.front();
+    
+    // Determine starting index of extra tuple vars
+    int numTempVars = ufcallmap.numTempVars();
+    int expandedArity = conj->arity();
+    int indexStart = expandedArity - numTempVars;
+//std::cout<<std::endl<<"St Idx = "<<indexStart<<"   Ex Ar = "<<expandedArity<<"   No Tm = "<<numTempVars<<std::endl<<std::endl;
+
+
+    SubMap affineSubstMap;
+    // Taking into account projection effect on tuple variable order:
+    // tv_n -> tv_n-1;  tv_n-1 -> tv_n-2; ... tv_tvar+1 -> tv_tvar
+    for (int i = tvar; i < expandedArity; i++) {
```

Comment:
We don't want to add new tuple vars for the uninterpreted functions, we want to add parameter variables (might be called symbolic constants in this code).  This will mean something different than UFCallsToTempVars.